### PR TITLE
[DNS Capture]: Do not allocate ip address for DNSLB service

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Fix part of #29224

Without this, 

if we have defined this SE:

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: external-svc-https
spec:
  hosts:
  - www.google.com
  location: MESH_EXTERNAL
  ports:
  - number: 443
    name: https
    protocol: TLS
  resolution: DNS

```

Then we could not access `curl http://www.google.com:80`, the address is unaccessible.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by ramaraochavali** [comment]
LGTM. Will leave it for @howardjohn to have look

**Review by howardjohn — COMMENTED** [review]
Don't we still want this? For example, see https://preliminary.istio.io/latest/blog/2020/dns-proxy/#external-tcp-services-without-vips

What we want is the app to resolve `mysql–instance1.us-east-1.rds.amazonaws.com` to auto allocated IP, but envoy to resolve it to mysql–instance1.us-east-1.rds.amazonaws.com upstream.

In the old implementation with envoy providing DNS, this worked out of the box, since it would go directly upstream.

Note that if we just change envoy to not get captured, it will break some other use cases (for example, define DNS SE for `foo.default.svc.cluster.local`, it will work on pods but not on VM). I think the ideal behavior would be:
```go
if fromApp {
  return autoAllocatedIP
} else {
  return realIpFromNDS.orElse(resolveUpstream)
}
```

File: pilot/pkg/serviceregistry/serviceentry/servicediscovery_test.go [hunk 1]
```diff
@@ -1313,10 +1313,9 @@ func Test_autoAllocateIP_conditions(t *testing.T) {
 			},
 			wantServices: []*model.Service{
 				{
-					Hostname:             "foo.com",
-					Resolution:           model.DNSLB,
-					Address:              "0.0.0.0",
-					AutoAllocatedAddress: "240.240.0.1",
```
Comment (kyessenov) [inline pilot/pkg/serviceregistry/serviceentry/servicediscovery_test.go:1319]: I'm curious: is there a chance of collision with real IP addresses here and breaking security policies over IPs (network or authz)?

**Comment by howardjohn** [comment]
240 is currently unassigned: https://ipgeolocation.io/resources/bogon.html.

**Comment by kyessenov** [comment]
@howardjohn I know, these are class E. I'm just wondering on implications on using new IPs with IP-based security policies. E.g. if a policy should deny based on IP, it seems it will behave differently depending on whether it uses a fake IP or not.

**Comment by howardjohn** [comment]
I see, thanks. Aren't all the policies applied on the server sidecar which
shouldn't see these fake IPs? I would expect they are only on the client
sidecar

**Comment by kyessenov** [comment]
I cannot confidently say one way or another. Perhaps, someone from the
security team can evaluate it.

**Comment by hzxuzhonghu** [comment]
>What we want is the app to resolve mysql–instance1.us-east-1.rds.amazonaws.com to auto allocated IP, but envoy to resolve it to mysql–instance1.us-east-1.rds.amazonaws.com upstream.

Yeah, IC. But currently, the tricky part is that for dns resolve, we could not control with port. 
This means for `www.google.com`, when app resolves `www.google.com`, it always resolved to 240.240.0.1, but if the port not defined, they it will be forwarded to the passthrough cluster and envoy will forward directly to `240.240.0.1`, which is unaccessible.

**Comment by hzxuzhonghu** [comment]
>Note that if we just change envoy to not get captured, it will break some other use cases (for example, define DNS SE for foo.default.svc.cluster.local, it will work on pods but not on VM).

@howardjohn How about only allocating ip for k8s service `foo.default.svc.cluster.local`, for others, leave it to upstream dns?

**Comment by howardjohn** [comment]
I think that doesn't really solve the problem? With k8s service we don't
need to auto allocate a fake ip, we already have a cluster IP.

**Comment by hzxuzhonghu** [comment]
I should say this is only for SE, i will file a commit first

**Comment by hzxuzhonghu** [comment]
Found another issue with 1.8.0, i have deployed httpbin application. And then after applying 
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: external-svc-https
spec:
  hosts:
  - httpbin.default.svc.cluster.local
  location: MESH_EXTERNAL
  ports:
  - number: 8000
    name: http
    protocol: HTTP
  resolution: DNS

```

Then the router generated is below:

```
 istioctl pc route sleep-8f795f47d-t5ft5 --name=8000 -ojson
[
    {
        "name": "8000",
        "virtualHosts": [
            {
                "name": "allow_any",
                "domains": [
                    "*"
                ],
                "routes": [
                    {
                        "name": "allow_any",
                        "match": {
                            "prefix": "/"
                        },
                        "route": {
                            "cluster": "PassthroughCluster",
                            "timeout": "0s",
                            "maxStreamDuration": {
                                "maxStreamDuration": "0s"
                            }
                        }
                    }
                ],
                "includeRequestAttemptCount": true
            },
            {
                "name": "httpbin.default.svc.cluster.local:8000",
                "domains": [
                    "httpbin.default.svc.cluster.local",
                    "httpbin.default.svc.cluster.local:8000",
                    "httpbin",
                    "httpbin:8000",
                    "httpbin.default.svc.cluster",
                    "httpbin.default.svc.cluster:8000",
                    "httpbin.default.svc",
                    "httpbin.default.svc:8000",
                    "httpbin.default",
                    "httpbin.default:8000",
                    "240.240.0.1",
                    "240.240.0.1:8000"
                ],
                "routes": [
                    {
                        "match": {
                            "prefix": "/xu",
                            "caseSensitive": true
                        },
                        "route": {
                            "weightedClusters": {
                                "clusters": [
                                    {
                                        "name": "outbound|8000||httpbin.default.svc.cluster.local",
                                        "weight": 85
                                    },
                                    {
                                        "name": "outbound|8000||sleep.default.svc.cluster.local",
                                        "weight": 15
                                    }
                                ]
                            },
                            "timeout": "3s",
                            "retryPolicy": {
                                "retryOn": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
                                "numRetries": 2,
                                "retryHostPredicate": [
                                    {
                                        "name": "envoy.retry_host_predicates.previous_hosts"
                                    }
                                ],
                                "hostSelectionRetryMaxAttempts": "5",
                                "retriableStatusCodes": [
                                    503
                                ]
                            },
                            "maxStreamDuration": {
                                "maxStreamDuration": "3s"
                            }
                        },
                        "metadata": {
                            "filterMetadata": {
                                "istio": {
                                    "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/foo"
                                }
                            }
                        },
                        "decorator": {
                            "operation": "foo:8000/xu*"
                        }
                    }
                ],
                "includeRequestAttemptCount": true
            }
        ],
        "validateClusters": false
    }
]


```

**Comment by hzxuzhonghu** [comment]
More findings: with the above se applied, we can not even access httpbin service, as there is no endpoints for the httpbin cluster, as we have two same services in istio. One is from k8s with resolution to ClientsideLB. and the other is from SE store with resolution DNS.  I can see the generated cluster is of type eds, but when generating eds, it get the service with DNS resolution and istio just skips this procedure, so no endpoints at all.

**Comment by howardjohn** [comment]
I opened https://github.com/istio/istio/issues/29324, ptal.

**Comment by hzxuzhonghu** [comment]
This is still necessary.

**Pull request closed without merging**